### PR TITLE
Allow a sorting function on min/max/min_by/max_by

### PR DESCRIPTION
**Pull request by josevalim**
Prior to this PR, there was no convenient way of getting the earliest/latest calendar type nor get the min/max datatype driven by a calendar type.

This PR adds this functionality by allowing a sorter function in `min/max/min_by/max_by` and allowing a sorter function to be automatically derived with a module by using `compare/2`, similar to #9426 

This PR introduces, however, a future deprecation, since the empty callback has been moved one argument ahead. This deprecation will be emitted only in the future (if we choose to do so).

### CODE REVIEW

**Comment by josevalim** [comment]
Tests added. This should be good to go.

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -1453,51 +1453,74 @@ defmodule Enum do
...
+  @spec max(t, (element, element -> boolean) | module(), (() -> empty_result)) ::
+          element | empty_result
+        when empty_result: any
+  def max(enumerable, sorter \\ &>=/2, empty_fallback \\ fn -> raise Enum.EmptyError end) do
+    {sorter, empty_fallback} = max_sort_fun(sorter, empty_fallback)
+    aggregate(enumerable, sorter, empty_fallback)
   end
 
+  # TODO: Deprecate me on 1.14
+  defp max_sort_fun(empty_fallback, default_empty_callback)
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]: There are other occurrences in the lines below.

```suggestion
  defp max_sort_fun(empty_fallback, default_empty_fallback)
```

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -1510,25 +1533,40 @@ defmodule Enum do
       iex> Enum.max_by(["a", "aa", "aaa", "b", "bbb"], &String.length/1)
       "aaa"
 
+  The fact this function uses Erlang's term ordering means that the
+  comparison is structural and not semantic. Therefore, if you want
+  to compare structs, most structs provide a "compare" function, such as
+  `Date.compare/2`, which receives two structs and returns `:lt` (less than),
+  `:eq` (equal), and `:gt` (greather than). If you pass a module as the
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  `:eq` (equal), and `:gt` (greater than). If you pass a module as the
```

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -1453,51 +1453,74 @@ defmodule Enum do
...
-  because the structural comparison compares the day before the year. This
-  can be addressed by using `max_by/3` and by relying on structures where
-  the most significant digits come first. In this particular case, we can
-  use `Date.to_erl/1` to get a tuple representation with year, month and day
-  fields:
+  In the example above, `max/2` returned April 1st instead of March 31st
+  because the structural comparison compares the day before the year.
+  For this reason, most structs provide a "compare" function, such as
+  `Date.compare/2`, which receives two structs and returns `:lt` (less than),
+  `:eq` (equal), and `:gt` (greather than). If you pass a module as the
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  `:eq` (equal), and `:gt` (greater than). If you pass a module as the
```

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -1510,25 +1533,40 @@ defmodule Enum do
...
+      iex> users = [
+      ...>   %{name: "Ellis", birthday: ~D[1943-05-11]},
+      ...>   %{name: "Lovelace", birthday: ~D[1815-12-10]},
+      ...>   %{name: "Turing", birthday: ~D[1912-06-23]}
+      ...> ]
+      iex> Enum.max_by(users, &(&1.birthday), Date)
+      %{name: "Ellis", birthday: ~D[1943-05-11]}
+
+  Finally, if you don't want to raise on empty enumerables, you can pass
+  the empty callback:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  the empty fallback:
```

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -1453,51 +1453,74 @@ defmodule Enum do
...
+  sorting function, Elixir will automatically use the `compare/2` function
+  of said module:
 
-      iex> Enum.max_by([~D[2017-03-31], ~D[2017-04-01]], &Date.to_erl/1)
+      iex> Enum.max([~D[2017-03-31], ~D[2017-04-01]], Date)
       ~D[2017-04-01]
 
-  For selecting a maximum value out of two consider using `Kernel.max/2`.
+  Finally, if you don't want to raise on empty enumerables, you can pass
+  the empty callback:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  the empty fallback:
```

File: lib/elixir/lib/enum.ex [hunk 6]
```diff
@@ -1453,51 +1453,74 @@ defmodule Enum do
...
       iex> Enum.max([~D[2017-03-31], ~D[2017-04-01]])
       ~D[2017-03-31]
 
-  In the example above, `max/1` returned March 31st instead of April 1st
-  because the structural comparison compares the day before the year. This
-  can be addressed by using `max_by/3` and by relying on structures where
-  the most significant digits come first. In this particular case, we can
-  use `Date.to_erl/1` to get a tuple representation with year, month and day
-  fields:
+  In the example above, `max/2` returned April 1st instead of March 31st
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  In the example above, `max/2` returned March 31st instead of April 1st 
```

File: lib/elixir/lib/enum.ex [hunk 7]
```diff
@@ -1575,51 +1613,74 @@ defmodule Enum do
...
-  because the structural comparison compares the day before the year. This
-  can be addressed by using `min_by/3` and by relying on structures where
-  the most significant digits come first. In this particular case, we can
-  use `Date.to_erl/1` to get a tuple representation with year, month and day
-  fields:
+  In the example above, `min/2` returned April 1st instead of March 31st
+  because the structural comparison compares the day before the year.
+  For this reason, most structs provide a "compare" function, such as
+  `Date.compare/2`, which receives two structs and returns `:lt` (less than),
+  `:eq` (equal), and `:gt` (greather than). If you pass a module as the
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  `:eq` (equal), and `:gt` (greater than). If you pass a module as the
```

File: lib/elixir/lib/enum.ex [hunk 8]
```diff
@@ -1575,51 +1613,74 @@ defmodule Enum do
...
+  sorting function, Elixir will automatically use the `compare/2` function
+  of said module:
 
-      iex> Enum.min_by([~D[2017-03-31], ~D[2017-04-01]], &Date.to_erl/1)
+      iex> Enum.min([~D[2017-03-31], ~D[2017-04-01]], Date)
       ~D[2017-03-31]
 
-  For selecting a minimal value out of two consider using `Kernel.min/2`.
+  Finally, if you don't want to raise on empty enumerables, you can pass
+  the empty callback:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  the empty fallback:
```

File: lib/elixir/lib/enum.ex [hunk 9]
```diff
@@ -1575,51 +1613,74 @@ defmodule Enum do
...
+  @spec min(t, (element, element -> boolean) | module(), (() -> empty_result)) ::
+          element | empty_result
+        when empty_result: any
+  def min(enumerable, sorter \\ &<=/2, empty_fallback \\ fn -> raise Enum.EmptyError end) do
+    {sorter, empty_fallback} = min_sort_fun(sorter, empty_fallback)
+    aggregate(enumerable, sorter, empty_fallback)
   end
 
+  # TODO: Deprecate me on 1.14
+  defp min_sort_fun(empty_fallback, default_empty_callback)
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]: Same here, and the lines below.

```suggestion
  defp min_sort_fun(empty_fallback, default_empty_fallback)
```

File: lib/elixir/lib/enum.ex [hunk 10]
```diff
@@ -1632,25 +1693,35 @@ defmodule Enum do
       iex> Enum.min_by(["a", "aa", "aaa", "b", "bbb"], &String.length/1)
       "a"
 
+  The fact this function uses Erlang's term ordering means that the
+  comparison is structural and not semantic. Therefore, if you want
+  to compare structs, most structs provide a "compare" function, such as
+  `Date.compare/2`, which receives two structs and returns `:lt` (less than),
+  `:eq` (equal), and `:gt` (greather than). If you pass a module as the
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  `:eq` (equal), and `:gt` (greater than). If you pass a module as the
```

File: lib/elixir/lib/enum.ex [hunk 11]
```diff
@@ -1632,25 +1693,35 @@ defmodule Enum do
...
+      iex> users = [
+      ...>   %{name: "Ellis", birthday: ~D[1943-05-11]},
+      ...>   %{name: "Lovelace", birthday: ~D[1815-12-10]},
+      ...>   %{name: "Turing", birthday: ~D[1912-06-23]}
+      ...> ]
+      iex> Enum.min_by(users, &(&1.birthday), Date)
+      %{name: "Lovelace", birthday: ~D[1815-12-10]}
+
+  Finally, if you don't want to raise on empty enumerables, you can pass
+  the empty callback:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  the empty fallback:
```

File: lib/elixir/lib/enum.ex [hunk 12]
```diff
@@ -1681,15 +1752,15 @@ defmodule Enum do
   end
 
   def min_max(enumerable, empty_fallback) when is_function(empty_fallback, 0) do
-    first_fun = &{&1, &1}
+    first_fun = &[&1 | &1]
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]: Why this change? (to learn 🙂)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by fuelen** [comment]
@josevalim What about Enum.min_max and Enum.min_max_by? Was this feature not implemented intentionally?

**Comment by josevalim** [comment]
@fuelen I remember briefly looking into this and I believe the issue is that we would need two functions, so I decided to not further explore it.
